lib/optimize/ConcatenatedModule.js

```diff
@@ -0,0 +1,489 @@
+				if(match) {
+					const referencedModule = modulesWithInfo[+match[1]];
+					let exportName;
+					if(match[2] === "ns") {
+						exportName = true;
+					} else {
+						const exportIdx = +match[2];
+						exportName = referencedModule.module.providedExports[exportIdx];
+					}
+					const finalName = getFinalName(referencedModule, exportName, moduleToInfoMap, requestShortener);
+					if(!finalName)
+						throw new Error(`Cannot map to variable name in module ${info.module.resource} (export '${exportName}')`);
+					const r = reference.identifier.range;
+					const source = info.source;
+					source.replace(r[0], r[1] - 1, finalName);
+				}
+			});
+		});
+
+		const result = new ConcatSource();
+		if(moduleToInfoMap.get(this.rootModule).needCompatibilityFlag) {
+			result.add(`Object.defineProperty(${this.rootModule.moduleArgument || "exports"}, "__esModule", { value: true });\n`);
+		}
+		modulesWithInfo.forEach(info => {
+			result.add(`\n// CONCATENAMED MODULE: ${info.module.readableIdentifier(requestShortener)}\n`);
+			if(info.needNamespaceObject) {
+				const name = info.exportMap.get(true);
+				const nsObj = [`var ${name} = {};`];
+				for(const exportName of info.module.providedExports) {
+					const finalName = getFinalName(info, exportName, moduleToInfoMap, requestShortener);
+					nsObj.push(`__webpack_require__.d(${name}, ${JSON.stringify(exportName)}, function() { return ${finalName}; });`);
+				}
+				result.add(nsObj.join("\n") + "\n");
+			}
+			result.add(info.source);
+		});
+
+		return result;
+	}
+
+	findNewName(oldName, usedNamed, extraInfo) {
+		let name = oldName;
+
+		if(name === "__WEBPACK_MODULE_DEFAULT_EXPORT__")
+			name = "defaultExport";
+
+		const splittedInfo = extraInfo.split("/");
+		while(splittedInfo.length) {
+			name = splittedInfo.pop() + "_" + name;
+			if(!usedNamed.has(Template.toIdentifier(name))) return Template.toIdentifier(name);
+		}
+
+		while(!usedNamed.has(name = name + "_"));
+		return name;
+	}
+
+}
+
+class HarmonyImportSpecifierDependencyConcatenatedTemplate {
+	constructor(originalTemplate, modulesMap) {
+		this.originalTemplate = originalTemplate;
+		this.modulesMap = modulesMap;
+	}
+
+	apply(dep, source, outputOptions, requestShortener, dependencyTemplates) {
+		const module = dep.importDependency.module;
+		const info = this.modulesMap.get(module);
+		if(!info) {
+			this.originalTemplate.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
+			return;
+		}
+		if(!Array.isArray(module.providedExports))
+			throw new Error(`Module ${module.resource} has no static exports ${module.providedExports}`);
+		let content;
+		if(dep.id === null) {
+			content = `__WEBPACK_MODULE_REFERENCE__${info.index}_ns__`;
+		} else {
+			const exportIdx = (module.providedExports).indexOf(dep.id);
+			content = exportIdx === -1 ? "undefined" : `__WEBPACK_MODULE_REFERENCE__${info.index}_${exportIdx}__`;
+		}
+		if(dep.shorthand) {
```

Comment:
Where is this set?

What is this intended to be used for; reexports inside a namespace object?